### ISSUE TITLE
Extend mpicc to include PMIx headers

### ISSUE THREAD

**Issue by rhc54** [open-mpi/ompi#10315]
We have a set of users (e.g., workflow managers) that need to directly access PMIx functions from within MPI applications. They are using OMPI as their MPI. Now that OMPI (starting with v5) directly uses PMIx itself, the PMIx functions are actually visible from within an OMPI program.

However, IIRC `mpicc` doesn't include `-I<path-to-pmix-headers>`, which means that apps wanting to access the PMIx functions cannot use `mpicc` to compile. This seems an unfortunate complication.

Would you be willing to add the necessary include paths to the `mpicc` command?

**Comment by ggouaillardet** [comment]
@rhc54 do you mean we should either add the path to the external PMIx headers that were used to build Open MPI or to the internal PMIx (if needed)?

From a packaging point of view, that would typically means `openmpi-devel` needs `pmix-devel`.
That would be a bit counterintuitive imho, but that's not a deal breaker.

Last but not least, we would have to add -L... -lpmix since some picky linkers might not allow the app to call PMIx subroutines.

**Comment by bwbarrett** [comment]
I'm not sure there's a great way to have mpicc include the PMIx paths by default, and we don't have a mechanism for adding options that's not a mess today.  If we're going to add the `-I`s for PMIx, that only makes sense if we also add the `-lpmix`, but now we've created a dependency for on applications that didn't exist before (today, applications only link against `-lmpi`.  While Open MPI no longer munges PMIx symbols or anything like that, since it doesn't explicitly link applications against `-lpmix`, customer can't portably make PMIx calls without doing the extra work to add the `-lpmix`.

We do the right things with the pkg-config modules.  In the meantime, setting `OMPI_CC` to `pmixcc` might be the best path forward for customers in this mode?

**Comment by rhc54** [comment]
> We do the right things with the pkg-config modules. In the meantime, setting `OMPI_CC` to `pmixcc` might be the best path forward for customers in this mode?

Sadly, I don't think so as `pmixcc` has no linkage to MPI and thus cannot compile an MPI application - there are no `-I`s or anything else for the MPI libraries. Nor is there a way to add the MPI flags as PMIx isn't configured with anything that would point to an MPI installation.

FWIW, this seems to be a use-case that is gathering steam. There is a EuroScience project specifically aimed at not only encouraging it, but creating the backend infrastructure in the RM to support it. They are using OMPI as their MPI, but this issue is causing some significant pain.

Since you are concerned about the added dependency for apps that don't need it (which I grok), would it make sense to perhaps create a new wrapper compiler that includes the necessary PMIx things? OMPI does know the necessary info via the configure logic to make the connection, so I would think it is possible.

**Comment by ggouaillardet** [comment]
I think @bwbarrett point is the end user can do
```
export OMPI_CC=pmixcc
mpicc foo.c
```

`mpicc` will add the paths/libs for Open MPI, and invoke `pmixcc` (instead of typically `gcc`) that will in turn adds the paths/libs for PMIx

does that make sense?
if yes, does it meet the needs of EuroScience?

**Comment by rhc54** [comment]
Ah, I didn't grok that trick. At first glance, that might work. However, it has one major flaw - it doesn't guarantee that the PMIx used by OMPI matches the PMIx being invoked by `pmixcc`. As we've seen with libevent and friends, having two different versions of the same lib being used by the app causes mucho badness.

What we need is a way to ensure that the MPI library and the app are both being built and linked against the _same_ PMIx install. The only way I can see to do that is to use OMPI's knowledge of which PMIx it built against. If we were adding it to `mpicc`, this would be a five minute fix - all you'd have to do is add an `OPAL_WRAPPER_ADD_xxx` where you already add the PMIx info to `CPPFLAGS`, `LDFLAGS`, and `LIBS`.

If we want to do it in a separate wrapper (e.g., `mpixcc`?), then it will take a little more effort as we'd have to duplicate `mpicc`, create a new set of "wrapper_add" macros, etc. Still, probably not more than a couple of hours to do.

**Comment by jsquyres** [comment]
(Note: after typing out this reply, I find it amusing that I'm channeling my inner Ralph in this reply: "don't make it complex!" 😜 )

I guess I'm not sure I understand the need for this complexity (new configure options, new flags, new wrapper compilers, etc.) because the building blocks are there for users already:

1. Use the nested wrapper compiler approach (e.g., setting `OMPI_CC` to `/path/to/pmixcc`), or
2. Use `pkg-config` files (in a few different ways) to get all the relevant flags

I guess I had always envisioned the work that we did for both of those things was so that users could do _exactly_ what is being asked for in this issue.  The intent was to give users friendly-to-the-rest-of-the-ecosystem solutions, rather than us inventing and maintaining another OMPI-specific solution.

We could even add a section about exactly this kind of issue to docs.open-mpi.org (somewhere in https://docs.open-mpi.org/en/v5.0.x/building-apps/index.html).

It is true that with both of the above options, the user has to make sure to select a correct/the same PMIx that Open MPI is using.  But that's very do-able, and is far less complex than a bunch of new m4 / OMPI infrastructure.

Just my $0.02...

**Comment by rhc54** [comment]
> It is true that with both of the above options, the user has to make sure to select a correct/the same PMIx that Open MPI is using. But that's very do-able, and is far less complex than a bunch of new m4 / OMPI infrastructure.

Well....not really. It kinda works if OMPI was using an external PMIx install - the user has to use `ompi_info` to find out which PMIx was used by OMPI, and then set the paths accordingly. Little tedious, but maybe good enough until the use-case gains more traction.

However, it doesn't work at all if OMPI built against its _internal_ version of PMIx. The user may have access to system-level versions, but that doesn't help. I guess you could just put a hard requirement that this use-case mandates the use of external PMIx installations. Just seems you are making it harder for users to utilize OMPI in this scenario.

I can/will talk to Ken about possibly providing support via MPICH - I don't think the EuroScience folks really care which one they use, but had glommed onto OMPI because of its PMIx integration. Still, it might be an option since MPICH doesn't have embedded PMIx code, so perhaps that would be a simpler solution?

**Comment by rhc54** [comment]
>     Use pkg-config files (in a few different ways) to get all the relevant flags

Should have addressed this suggestion. Yes, and that is what the current developers are forced to do (or at least a variant of that suggestion). However, if you consider the average user, that's a pretty poor option. It puts a significant burden on a chemist or physicist to say you have to figure out pkg-config to compile your code, or use `mpicc --showme` and then edit/replace options to do it.

Unfortunately, as I said, layering the wrappers doesn't work in the general case, so that option isn't really tractable in a production environment. We are looking for a production-suitable solution, but perhaps that isn't something OMPI is prepared to provide right now (and I am not criticizing that thinking).

One option (that we are currently doing) is for us to simply generate a patch that does what we need, and for the Euro guys to redistribute the patched versions of OMPI. Kinda ugly, but it may be the only path forward if MPICH cannot (or at least until they can) fill the gap. I'm okay with generating the patch and letting them redistribute as required, but I'd rather not do that long term.

**Comment by rhc54** [comment]
> What about adding yet an other configure option (such as --enable-pmix-in-wrappers, "default" is false to avoid adding an other dependency for the package managers) that does the trick?

Sure, that would work - though I don't understand the dependency comment. OMPI already depends on PMIx, so I don't see what is being added from the point of view of the package manager?

**Comment by rhc54** [comment]
Fiddled around a bit and here is what I found.

When using the embedded PMIx:
```text
$ mpicc --showme
gcc -I/Users/rhc/openmpi/build/foobar/include -L/Users/rhc/openmpi/build/foobar/lib -lmpi
```
To compile a PMIx-based program, I only had to add `-lpmix` - the include path already covers `pmix.h` and friends.

When using an external PMIx:
```text
$ mpicc --showme
gcc -I/Users/rhc/openmpi/build/foobar/include -L/Users/rhc/openmpi/build/foobar/lib -lmpi
```
So the wrapper compiler remained the same, which means you have to add the other flags as well:
```text
$ gcc -I/Users/rhc/openmpi/build/foobar/include -L/Users/rhc/openmpi/build/foobar/lib -lmpi 
          -I/Users/rhc/pmix/build/master/include -L/Users/rhc/pmix/build/master/lib -lpmix
```
I'll work with the Euro guys to devise a path forward. For now, I can provide them with a patch they can distribute, and we can explore the MPICH path in parallel. They'll need to support MPICH at some point anyway, so it's more a question of which comes first, which one they use for the papers, etc. MPICH is in some ways easier as we don't have this question of what has to be added to the cmd line to make it work - there is only one configuration.

Let me know if you guys come up with a solution. You are welcome to close this for now if you prefer.

**Comment by ggouaillardet** [comment]
Currently, `openmpi-devel` (that provides `mpicc` and `libmpi.so`) depends on `openmpi` (for `mpirun` and `libmpi.so.0`) and `pmix` (for `libpmix.so.0`). With these changes, `openmpi-devel` would also have to depend on `pmix-devel` (for `libpmix.so` and `pmix.h`)

**Comment by rhc54** [comment]
The dependencies are a real can of worms, frankly. You neglected that `openmpi` actually has to depend on `prrte` since `mpirun` is just a wrapper over `prterun`.

Your `ompi-devel` package has to depend on `pmix-devel` since you directly call PMIx functions, so that dependency already exists.

PRRTE has been modified to use the PMIx utilities for code reduction, so `prrte-devel` now also requires `pmix-devel`. OMPI doesn't require `prrte-devel` to compile.

Bottom line therefore is that both `ompi-devel` and `prrte-devel` already depend on `pmix-devel`, and `pmix-devel` will always contain all the PMIx headers (not just `pmix.h` and friends) so that PRRTE can build. So I don't see how anything changes - did I miss something?

**Comment by jsquyres** [comment]
> Your `ompi-devel` package has to depend on `pmix-devel` since you directly call PMIx functions, so that dependency already exists.

I'm not sure I understand this comment:

* I think that what is being referred to here as `ompi-devel` is the package including the OMPI wrapper compilers (e.g., `mpicc`) and the OMPI header files.  I'm assuming that the intent for the `ompi-devel` package is to compile MPI applications (not the core Open MPI libraries).
* I don't know what precisely is implied by `pmix-devel`, but I would imagine it is the PMIx wrapper compilers and the PMIx header files, and its intent is to compile PMIx applications (not the core PMIx libraries).

Today:

* To compile a simple MPI application (e.g., https://github.com/open-mpi/ompi/blob/main/examples/hello_c.c), you do not need a PMIx wrapper compiler or the PMIx header files.
* The core Open MPI libraries depend on libpmix.  
* MPI applications do not depend directly on libpmix; it is indirectly pulled in by the core Open MPI libraries.

If we make this change, then `mpicc` will always require the PMIx headers.  That's new, and gives me pause.

Per above discussion, I thought one of the major design points of this whole infrastructure was that PMIx was getting to be ubiquitous outside of the bundled version in Open MPI.  It's inbox in distros, available in homebrew and other package managers, and people are even downloading it independently to get the most recent version (outside of Open MPI).  Hence, I thought the whole point was that the Open MPI-bundled PMIx would rarely be used.  In such a case, I guess I don't understand why the focus is back on the Open MPI-bundled PMIx...?  Isn't community-norms like `pkg-config` and / or setting a single env variable to use a nested wrapper compiler sufficient for what we are hoping/assuming is the case (that PMIx is external to Open MPI)?

**Comment by ggouaillardet** [comment]
> Your ompi-devel package has to depend on pmix-devel since you directly call PMIx functions, so that dependency already exists.

Open MPI directly calls PMIx functions, and that means `pmix-devel` is required in order to **build** Open MPI.
I do not see how that makes `pmix-devel` a runtime requirement for `openmpi-devel`

**Comment by rhc54** [comment]
The problem is that OMPI continues to include internal versions of the package, so the user cannot readily know whether the library they are referencing uses the internal or an external version of PMIx, and if the latter, then which external version was used. Yes, they could be trained to use `ompi_info` or some other mechanism to figure it all out - but again, this generates extra steps that create barriers.

Nobody is asking you to hardware `mpicc` to include the necessary PMIx references. All we were asking for was to add a `--with-pmix` option that would add the references to the wrapper compiler command. Thus, if I _want_ to use PMIx directly from within an OMPI-built program, I can add a simple option that ensures the compatible PMIx library is being used and has been added.

Seems pretty trivial - frankly, I'm surprised by the kerfluffle. It was only a few line patch, took a handful of minutes to do.

As for the packages, yes I confused what @ggouaillardet was saying. However, that seems a pretty simple thing for packagers to document, and admins can easily install the `pmix-devel` package if users request that capability. I guess I just don't understand why that is an issue? If the package isn't installed, then the compile option will fail as the headers won't be found - and someone can ask that the package be installed. Am I missing something?

**Comment by rhc54** [comment]
Just got burned by this again while helping Argonne - nice segfaults due to conflicting PMIx libraries being picked up. Made worse by apparent bugs in the OMPI configure code that are blocking the ability to build with external PMIx/PRRTE libraries.

Sigh

Argonne will file the configure bugs - if you truly object to providing the `mpicc` option, I guess I'll have to find a way to distribute the changes. This person is highly knowledgeable, but it was super kludgy to figure out how to make this work and took quite a bit of our time.

**Comment by rhc54** [comment]
Well, the EuroScience project switched to MPICH, and I gather others are going the same way when wanting to use PMIx in their MPI apps. Easier than me trying to circulate and maintain an OMPI patch. So I guess we can close this - people wanting to use PMIx in their apps will just be referred to MPICH.
